### PR TITLE
errors: simplify referring to fluent attributes

### PR DESCRIPTION
**Pull request by davidtwco**
To render the message of a Fluent attribute, the identifier of the Fluent message must be known. `DiagnosticMessage::FluentIdentifier` contains both the message's identifier and optionally the identifier of an attribute. Generated constants for each attribute would therefore need to be named uniquely (amongst all error messages) or be able to refer to only the attribute identifier which will be combined with a message identifier later. In this commit, the latter strategy is implemented as part of the `Diagnostic` type's functions for adding subdiagnostics of various kinds.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_macros/src/diagnostics/fluent.rs [hunk 1]
```diff
@@ -199,15 +203,19 @@ pub(crate) fn fluent_messages(input: proc_macro::TokenStream) -> proc_macro::Tok
                 });
 
                 for Attribute { id: Identifier { name: attr_name }, .. } in attributes {
-                    let attr_snake_name = attr_name.replace("-", "_");
-                    let snake_name = Ident::new(&format!("{snake_name}_{attr_snake_name}"), span);
+                    let snake_name = Ident::new(&attr_name.replace("-", "_"), span);
+                    if previous_attrs.contains(&snake_name) {
```
Comment (oli-obk) [inline compiler/rustc_macros/src/diagnostics/fluent.rs:207]:
```suggestion
                    if !previous_attrs.insert(&snake_name) {
```

File: compiler/rustc_error_messages/src/lib.rs [hunk 2]
```diff
@@ -252,6 +294,29 @@ pub enum DiagnosticMessage {
...
+            SubdiagnosticMessage::Str(s) => return DiagnosticMessage::Str(s.clone()),
+            SubdiagnosticMessage::FluentIdentifier(id) => {
+                return DiagnosticMessage::FluentIdentifier(id, None);
+            }
+            SubdiagnosticMessage::FluentAttr(attr) => attr,
+        };
+
+        match self {
+            DiagnosticMessage::Str(s) => DiagnosticMessage::Str(s.clone()),
+            DiagnosticMessage::FluentIdentifier(id, _) => {
```
Comment (oli-obk) [inline compiler/rustc_error_messages/src/lib.rs:314]: Does this ever happen where then `_` is `Some` already?

**Review by oli-obk — APPROVED** [review]
just a nit and a question, r=me with those resolved/answered

File: compiler/rustc_error_messages/src/lib.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_error_messages/src/lib.rs:314]: It shouldn't as the code is written right now, but it wouldn't be much of a problem if it did. I considered adding a panic or something like that.

File: compiler/rustc_macros/src/diagnostics/fluent.rs (hunk 1 above)
Comment (davidtwco) [reply compiler/rustc_macros/src/diagnostics/fluent.rs]: Fixed!

**Comment by pnkfelix** [comment]
visiting for rustc-perf performance triage.

There were some regressions here, but they are few, minor, and contained solely to secondary benchmarks (specifically projection-caching and wg-grammar). Marking as triaged.

@rustbot label: perf-regression-triaged

**Pull request merged**